cpp/oneapi/dal/algo/jaccard/backend/cpu/vertex_similarity_default_kernel_skx.cpp

```diff
@@ -31,17 +30,688 @@ namespace oneapi::dal::preview {
 namespace jaccard {
 namespace detail {
 
-template vertex_similarity_result
-call_jaccard_default_kernel_avx2<oneapi::dal::backend::cpu_dispatch_avx512>(
-    const descriptor_base &desc,
-    vertex_similarity_input<undirected_adjacency_array_graph<>> &input);
+#if defined(__INTEL_COMPILER)
+DAAL_FORCEINLINE std::int32_t _popcnt32_redef(const std::int32_t &x) {
+    return _popcnt32(x);
+}
+#define GRAPH_STACK_ALING(x) __declspec(align(x))
+#else
+DAAL_FORCEINLINE std::int32_t _popcnt32_redef(const std::int32_t &x) {
+    std::int32_t count = 0;
+    std::int32_t a = x;
+    while (a != 0) {
+        a = a & (a - 1);
+        count++;
+    }
+    return count;
+}
+#define GRAPH_STACK_ALING(x)
+#endif
+
+DAAL_FORCEINLINE std::size_t intersection(std::int32_t *neigh_u,
+                                          std::int32_t *neigh_v,
+                                          std::int32_t n_u,
+                                          std::int32_t n_v) {
+    size_t total = 0;
+    std::int32_t i_u = 0, i_v = 0;
+    while (i_u < (n_u / 16) * 16 && i_v < (n_v / 16) * 16) { // not in last n%16 elements
+
+        // assumes neighbor list is ordered
+        std::int32_t minu = neigh_u[i_u];
+        std::int32_t maxv = neigh_v[i_v + 15];
+
+        if (minu > maxv) {
+            if (minu > neigh_v[n_v - 1]) {
+                return total;
+            }
+            i_v += 16;
+            continue;
+        }
+
+        std::int32_t minv = neigh_v[i_v];
+        std::int32_t maxu = neigh_u[i_u + 15];
+        if (minv > maxu) {
+            if (minv > neigh_u[n_u - 1]) {
+                return total;
+            }
+            i_u += 16;
+            continue;
+        }
+        __m512i v_u = _mm512_loadu_si512((void *)(neigh_u + i_u)); // load 16 neighbors of u
+        __m512i v_v = _mm512_loadu_si512((void *)(neigh_v + i_v)); // load 16 neighbors of v
+        if (maxu >= maxv)
+            i_v += 16;
+        if (maxu <= maxv)
+            i_u += 16;
+
+        __mmask16 match = _mm512_cmpeq_epi32_mask(v_u, v_v);
+        if (_mm512_mask2int(match) != 0xffff) { // shortcut case where all neighbors match
+            __m512i circ1 = _mm512_set_epi32(0, 15, 14, 13, 12, 11, 10, 9, 8, 7, 6, 5, 4, 3, 2, 1);
```

Comment:
I suppose it will be faster to replace this:
```cpp
__m512i circ1 = _mm512_set_epi32(0, 15, 14, 13, 12, 11, 10, 9, 8, 7, 6, 5, 4, 3, 2, 1);
..
__m512i circ15 = _mm512_set_epi32(14, 13, 12, 11, 10, 9, 8, 7, 6, 5, 4, 3, 2, 1, 0, 15);

__m512i v_v1 = _mm512_permutexvar_epi32(circ1, v_v);
...
__m512i v_v15 = _mm512_permutexvar_epi32(circ15, v_v);
```
By:
```cpp
alignas(256) __m512i buff[2];
buff[0] =  v_v;
buff[1] =  v_v;

__m512i v_v1 = _mm512_loadu_si512((void *)(((int*)buff) + 0));
...
__m512i v_v1 =_mm512_loadu_si512((void *)(((int32_t*)buff) + 14));
...
```
I hope that reverse of element's order isn't important for you.

Throughput of ```_mm512_permutexvar_epi32``` is equal to 1 clocks/instruction (CPI), but ```_mm512_loadu_si512``` is equal to 0.5 CPI. Also, ```_mm512_set_epi32``` has undefined CPI because it's some sequence of instructions defined by a compiler.

Reply:
>>> I hope that reverse of element's order isn't important for you. 
No, it's important. We have to compare all the permutations of the set. 
Probably we can't apply the suggested change.

Reply:
Your idea is just iterate over all possible combinations and do comparisons. The order of the comparisons doesn't matter here.

Reply:
@SmirnovEgorRu ok, yes, you're right, got it! will check today

Reply:
> Your idea is just iterate over all possible combinations and do comparisons. The order of the comparisons doesn't matter here.

It should works. With alignas we can use _mm512_load_si512.

Reply:
@orrrrtem, for the first element - yes, for other - no, because we shift buff by some INT's.

Reply:
@SmirnovEgorRu checked this option - got ~1% improvement, so looks like on-par. 
But the code looks simpler, probably we can use this version.

Reply:
@SmirnovEgorRu, this algorithm is bounded by memory. The last part of our perf improvements was about memory. Any suggestions for prefetch parts?

Reply:
If the hotspot is memory - acceleration of the computation will not bring anything. Yes, it's clear. Probably for Icelake (with higher BW) computation portion will take more % of time.

About work with memory acceleration - need to profile, if the hotspot is memory BW, prefetch shouldn't help.